pkg/status/reporter.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/projectcalico/libcalico-go/lib/options"
+
+	log "github.com/sirupsen/logrus"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+)
+
+// Interface for a component to populate its status to node status resource.
+type statusPopulator interface {
+	Populate(status *apiv3.CalicoNodeStatus) error
+	Show()
+}
+
+// reporter contains all the data/method about reporting back node status based on a single node status resource.
+// Each reporter has a goroutine which constantly reads node status and updates node status resource.
+type reporter struct {
+	// The name of the node status resource.
+	name string
+
+	// Internal client to operate on node status resource.
+	client client.Interface
+
+	// buffered channel to receive updates for the resource.
+	ch chan *apiv3.CalicoNodeStatus
+
+	// status holds latest version of node status resource.
+	status *apiv3.CalicoNodeStatus
+
+	// Interval and Time ticker that node status should be reported.
+	interval int
+	ticker   *time.Ticker
+
+	// populators
+	populators map[BirdConnType]map[apiv3.NodeStatusClassType]statusPopulator
+
+	// channel to indicate this reporter is not needed anymore.
+	// It should start termination process.
+	done chan struct{}
+
+	// channel to indicate this reporter is terminated.
+	term chan struct{}
+}
+
+// newReporter creates a reporter and start running a goroutine handling resource update.
+func newReporter(name string, client client.Interface, intervalInSeconds int, populators map[BirdConnType]map[apiv3.NodeStatusClassType]statusPopulator) *reporter {
```

Comment:
I think maybe we should have `newReporter` take a NodeStatus object so that we can seed it with initial configuration before calling `run()`

I think it simplifies a bit of the logic, since we'll never have a reporter without a `status` set, which is an invalid state anyway really (since reporters only exist because of a status object)

Reply:
Good point.